File: ios/Classes/FlutterChannelEventBusMixin.swift
//
//  FlutterChannelEventBusMixin.swift
//  flutter_channel_event_bus
//
//  Created by joser on 2021/6/3.
//

import Foundation

public protocol FlutterChannelEventBusMixin {
    
}

fileprivate struct FlutterChannelEventBusMixinKey {
    static var _eventBusRegisterIds = "_eventBusRegisterIds"
}

public extension FlutterChannelEventBusMixin {
    private var _eventBusRegisterIds:[String] {
        let ids = objc_getAssociatedObject(self,
                                           &FlutterChannelEventBusMixinKey._eventBusRegisterIds)
        return ids as? [String] ?? []
    }
    /// 释放当前所有的注册订阅
    func freeRegisterIds() {
        self._eventBusRegisterIds.forEach { element in
            FlutterChannelEventBus.default.deregister(event: element)
        }
    }
    
    
    /// 通过默认的路由进行注册订阅
    /// - Parameters:
    ///   - name: 注册的订阅名称
    ///   - handle: 注册订阅消息回掉
    /// - Returns: 注册的订阅ID
    func register(eventBus name:String,
                  handle:@escaping FlutterChannelEventBus.FlutterChannelEventBusResponsehandle) -> String {
        let registerId = FlutterChannelEventBus.default.register(event: name,
                                                                 handle: handle)
        append(eventBus: registerId)
        return registerId
    }
    
    /// 通过自定义的路由进行注册订阅
    /// - Parameters:
    ///   - name: 注册的订阅名称
    ///   - routeName: 自定义路由的名称
    ///   - handle:  注册订阅消息回掉
    /// - Returns: 注册的订阅ID
    func register(customEventBus name:String,
                  routeName:String,
                  handle:@escaping FlutterChannelEventBus.FlutterChannelEventBusResponsehandle) -> String {
        let registerId = FlutterChannelEventBus.default.register(event: name,
                                                                 route: .init(custom: routeName),
                                                                 handle: handle)
        append(eventBus: registerId)
        return registerId
    }
    
    /// 通过默认的路由发送一条订阅消息
    /// - Parameters:
    ///   - name: 订阅的名称
    ///   - data: 消息的内容
    ///   - result: 接收到回掉
    func send(eventBus name:String,
              data:Any? = nil,
              result:@escaping FlutterResult) {
        FlutterChannelEventBus.default.send(event: name,
                                            data: data,
                                            result: result)
    }
    
    /// 通过自定义的路由发送一条订阅消息
    /// - Parameters:
    ///   - name: 订阅的名称
    ///   - routeName: 自定义路由名称
    ///   - data: 消息的内容
    ///   - result: 接收到回掉
    func send(customEventBus name:String,
              routeName:String,
              data:Any? = nil,
              result:@escaping FlutterResult) {
        FlutterChannelEventBus.default.send(event: name,
                                            route: .init(custom: routeName),
                                            data: data,
                                            result: result)
    }
    
    private func append(eventBus id:String) {
        var ids:[String] = self._eventBusRegisterIds
        ids.append(id)
        objc_setAssociatedObject(self,
                                 &FlutterChannelEventBusMixinKey._eventBusRegisterIds,
                                 ids,
                                 .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }
}

File: lib/flutter_channel_event_bus_mixin.dart
import 'package:flutter_channel_event_bus/flutter_channel_event_bus.dart';

mixin FlutterChannelEventBusMixin {
  /// 存放已经注册ID数组
  final List<String> _eventBusRegisterIds = [];

  /// 释放注册的通知
  void freeRegisters() {
    for (var element in _eventBusRegisterIds) {
      FlutterChannelEventBus.defaultEventBus.deregister(element);
    }
  }

  /// 注册一个通用的通知
  /// [name]注册的名称
  /// [handle]消息回掉
  String register(
    String name,
    FlutterChannelEventBusResponseHandle handle,
  ) {
    var route = FlutterChannelEventBusRoute.general();
    var registerId = FlutterChannelEventBus.defaultEventBus.register(
      name,
      route,
      handle,
    );
    _eventBusRegisterIds.add(registerId);
    return registerId;
  }

  /// 注册一个自定义的通知
  /// [name]注册的名称
  /// [routeName]自定义消息路由
  /// [handle]消息回掉
  String registerCustom(
    String name,
    String routeName,
    FlutterChannelEventBusResponseHandle handle,
  ) {
    var route = FlutterChannelEventBusRoute.custom(routeName);
    var registerId = FlutterChannelEventBus.defaultEventBus.register(
      name,
      route,
      handle,
    );
    _eventBusRegisterIds.add(registerId);
    return registerId;
  }

  /// 通过通用的路由发送消息
  /// [name]消息的名称
  /// [data]消息的数据
  Future<dynamic> send(
    String name,
    dynamic data,
  ) {
    var route = FlutterChannelEventBusRoute.general();
    return FlutterChannelEventBus.defaultEventBus.send(
      name,
      route,
      data,
    );
  }

  /// 通过自定义的路由发送消息
  /// [name]消息名称
  /// [routeName]自定义路由名称
  /// [data]消息的数据
  Future<dynamic> sendCustom(
    String name,
    String routeName,
    dynamic data,
  ) {
    var route = FlutterChannelEventBusRoute.custom(routeName);
    return FlutterChannelEventBus.defaultEventBus.send(
      name,
      route,
      data,
    );
  }
}
